STREAMS-1849 Provide at-least-once Quality of Service (QoS) level (#76)
Axway/streams-open-docs: content/en/docs/subscribers/subscriber-sse.md

description: Learn how to configure and use the Streams Server-Sent Events Subscriber.
---

> Server-Sent Events (SSE) is part of the HTML5 standard. SSEs are sent over traditional HTTP, so it doesn't require a special protocol to get working. SSE embedded important features such event identifier, automatic reconnection or the ability to manage heterogeneous events.

Reviewer: ```suggestion
Server-Sent Events (SSE) is part of the HTML5 standard. SSEs are sent over traditional HTTP, so they don't require a special protocol to get working. SSE embedded important features, such event identifier, automatically reconnect or the have the ability to manage heterogeneous events.
```

Author: on the last sentence, "the have the ability" it looks like a word is missing, isn't it?

Reviewer: Server-Sent Events (SSE) is part of the HTML5 standard. SSEs are sent over traditional HTTP, so they don't require a special protocol to work. SSE embedded important features, such event identifier, either automatically reconnect or they have the ability to manage heterogeneous events.

Author: Ok I fixed it.

> Server-Sent Events (SSE) is part of the HTML5 standard. SSEs are sent over traditional HTTP, so they don't require a special protocol to work. SSE embedded important features, such event identifier, either automatically reconnect or they have the ability to manage heterogeneous events.

Reviewer 2: "**SSE embedded important features, such event identifier...**" --> This is a bit confusing to me. Do you mean, "SSE includes important features, such as, `Last-Event-ID header` event identifier, which either automatically reconnects a broken connection or _can manage heterogeneous events_."

(...can manage heterogeneous events ???) - even with suggestion above, this last part is still not clear

Reviewer 3: @andreamussap Here my proposal: SSE includes important features, such as, `Last-Event-ID` header support, automatic reconnection and heterogeneous event handling.

These three fields are always present and represent a single message also called an event.

### Compression

SSE can be compressed on demand using gzip or deflate methods. The following is an example of how to use the `Accept-Encoding` header:

```sh
export BASE_URL="base-url"
export TOPIC_ID="topic-id"

curl -v "${BASE_URL}/streams/subscribers/sse/api/v1/topics/${TOPIC_ID}" -H "Accept-Encoding: gzip, deflate" --compress
```

If this header is not provided, the default behavior is not to compress the data.

### Last-Event-Id

SSE has the ability for clients to automatically reconnect if the connection is interrupted. Furthermore, the data stream continues from the point it disconnected, so no events are lost.

> Each message sent by Streams is uniquely identified. Using the built-in header `Last-Event-Id` after reconnecting, the client can tell Streams where in the events streams to resume. Such mechanism is automatically integrated in most clients but you can also achieve it using curl command, using the example given above it will be:

Reviewer: ```suggestion
Each message sent by Streams is uniquely identified. Using the built-in header `Last-Event-Id` after reconnecting, the client can tell Streams where in the events streams to resume. This mechanism is automatically integrated in most clients, but you can also achieve it using the curl command:
```

> Each messages sent by Streams are uniquely identified, so using the built-in header `Last-Event-id` after a reconnection, the client can tell to Streams where in the events streams to resume. Such mechanism is automaticaly integrated in most clients but you can also acheive it using curl command, using the example given above it will be :

Reviewer 3: ```suggestion
Each message sent by Streams is uniquely identified. Using the built-in header `Last-Event-Id` after reconnecting, the client can tell Streams where in the events streams to resume. Such mechanism is automatically integrated in most clients but you can also achieve it using curl command, using the example given above it will be:
```
